### PR TITLE
Element unique selectors for target and coded regions in the visual grid

### PR DESCRIPTION
**Pull request by amitzur**
This PR does a bit of order in the `toCheckWindowConfiguration` and `toPersistedRegions` area.
Points to note:
1. Before taking a DOM snapshot, all necessary elements are found and then marked.
2. After DOM snapshot they are cleaned up.
3. `toCheckWindowConfiguration` is now synchronous, and its job is just to map and merge checkSetting and configuration in the VGC's checkWindow input.
4. `toPersistedRegions` is now also synchronous, and its job is just to map from the element's unique ID to the selector structure.
5. `toEyesSelector` is no longer needed (and `getElementXpath` as well, but maybe we want to keep that)
6. I'm saving in a WeakMap the selectors for the elements.
7. This is a preliminary step before using `enableMultipleResultsPerSelector`. Once we do that, we can use just one GUID for all the elements that we want regions for. Or maybe not one, but one per type (ignore, floating, etc.)

### CODE REVIEW

**Comment by amitzur** [comment]
@kyryloonufriiev BTW, this also should fix the issue in protractor with ElementFinder, because I'm always working with EyesElement now.

**Review by kyryloonufriiev — COMMENTED** [review]
I like this PR for a couple of reasons.
- Remove `toEyesSelector` is a good move. Less code in `SpecDriver` is great.
- Execute max two snippets per entire check settings object is great as well.

But I would like to suggest something. Right now `elementIds` are generated separately from `FluentRegion` object. I think a better approach will be to generate id-element mapping during resolving elements and save it in `FluentRegion` object. Then we could just return ids when `toPersistedRegion` called without need to provide a mapping as an argument. The whole logic will be more consistent since the ids are an implementation detail and we don't care about them.
The only reason it couldn't be seamless is that we want to mark all elements in one snippet. WDYT?

File: packages/eyes-sdk-core/lib/fluent/CheckSettingsUtils.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    floating: persistRegions(checkSettings.getFloatingRegions()),
+    strict: persistRegions(checkSettings.getStrictRegions()),
+    layout: persistRegions(checkSettings.getLayoutRegions()),
+    content: persistRegions(checkSettings.getContentRegions()),
+    accessibility: persistRegions(checkSettings.getAccessibilityRegions()),
+    target: !checkSettings._targetRegion && !checkSettings._targetElement ? 'window' : 'region',
+    fully: TypeUtils.getOrDefault(
+      checkSettings.getStitchContent(),
+      configuration.getForceFullPageScreenshot(),
+    ),
```
[the remark this answers is not in the document]
Comment (amitzur) [reply packages/eyes-sdk-core/lib/fluent/CheckSettingsUtils.js:38]: You're right! I will fix it.
What's strange is that also `sendDom` behaves that way (configuration takes precedence over checkSettings). But I don't want to change it for users now. I should communicate it first. So I will fix line 41 as well.

File: packages/eyes-sdk-core/lib/sdk/EyesVisualGrid.js [hunk 2]
```diff
@@ -238,28 +240,37 @@ class EyesVisualGrid extends EyesCore {
...
     return this._checkPrepare(checkSettings, async () => {
-      // this._logger.verbose(`Dom extraction starting   (${checkSettings.toString()})   $$$$$$$$$$$$`)
-      const pageDomResults = await this.constructor.VisualGridClient.takeDomSnapshot({
-        executeScript: this._context.execute.bind(this._context),
-      })
-      const {cdt, url, resourceContents, resourceUrls, frames} = pageDomResults
-      if (this.getCorsIframeHandle() === CorsIframeHandles.BLANK) {
-        CorsIframeHandler.blankCorsIframeSrcOfCdt({url, cdt, frames})
+      const regionElements = await getAllRegionElements({checkSettings, context: this._driver})
+      debugger
```
[the remark this answers is not in the document]
Comment (amitzur) [reply packages/eyes-sdk-core/lib/sdk/EyesVisualGrid.js]: :+1: also added a no-debugger eslint rule

**Comment by amitzur** [comment]
> I think a better approach will be to generate id-element mapping during resolving elements and save it in `FluentRegion` object. 
> The only reason it couldn't be seamless is that we want to mark all elements in one snippet. WDYT?

That would be very good. Right now, in addition to the 2 snippet I run, there is a `context.elements` call for every region. I can avoid needing the `markElements` snippet, and replace `context.elements` with a snippet that finds and marks the elements. This way it'll all be handles inside `FluentRegion`. That's a great abstraction. Only thing I'm not use about is - would it be equivalent to run this snippet vs. `context.elements`?

Here's what I meant in code:

```
// FluentRegion
async resolveElements(context) {
  const arg = this._selector || this._element
  if (arg) {
    this._elementIds = await EyesUtils.findAndMarkElements(context, arg)
  }
}

toPersistedRegions() {
  if (this._region) {
    return [{...this._region.toJSON(), ...this._options}]
  } else {
    return this._elementIds.map(id => ({
      ...this._options,
      type: 'css',
      selector: `[data-eyes-selector="${id}"]`,
    }))
  }
}

// snippet
function findAndMarkElements(elementOrSelector) {
  const elements = typeof elOrSelector === 'string' ? document.querySelectorAll(elementOrSelector) : [elementOrSelector]
  return elements.map(el => {
    const id = `${Math.random()}`.substr(2)
    el.setAttribute('data-eyes-selector', id)
    return id
  })
}
```

**Comment by kyryloonufriiev** [comment]
@amitzur I don't think we need to do `findAndMarkElements` since selector could be not `css` (e.g. link text or xpath).
My idea was to avoid the need to handle by ourselves an `elementIds` mapping and also place it in a place that is related to this logic.

The algorithm I thought about is:
1. Call `resolveElements` (might be renamed since the logic was changed) for each `FluentRegion` object.
        - Find all elements if needed
        - Generate guid for each element, and save them in this object
        - Return a guid -> element mapping
2. Collect all elements in one array and execute `markElementIds` snippet
3. Call `getPersistedRegions` for each `FluentRegion` object.
        - Create a selector from each guid and return an array
4. Execute `clearElementIds` snippet.

**Comment by amitzur** [comment]
@kyryloonufriiev ah yes! perfect solution. I will do it.

**Pull request merged**